src/server.c

```diff
@@ -1256,8 +1256,7 @@ int serverCron(struct aeEventLoop *eventLoop, long long id, void *clientData) {
     /* Start a scheduled AOF rewrite if this was requested by the user while
      * a BGSAVE was in progress. */
     if (!hasActiveChildProcess() &&
-        server.aof_rewrite_scheduled &&
-        !aofRewriteLimited())
```

Comment:
@chenyang8094 looking at the code behind an unsuccessful initial rewwrite,
`startAppendOnly` creates a manifest with only an incr file, i think that the code at the bottom of `backgroundRewriteDoneHandler` (the one that sets aof_rewrite_scheduled to 1), should delete these files incremental file, and also the manifest)
otherwise, each retry of aof_rewrite_scheduled will add another incremental file.

am i missing anything?

Reply:
@oranagra  Oh，you are right.  

When we execute `config set appendonly yes` in a problematic system, it will indeed trigger unlimited AOFRW and quickly generate INCR AOF files. I remembered the original meaning of adding `aofRewriteLimited` to avoid generating so many INCR AOF files and reducing CPU waste as much as possible.

So, I think my previous suggestion is wrong and we should keep the limit here.

```
    if (!hasActiveChildProcess() &&
        server.aof_rewrite_scheduled &&
        !aofRewriteLimited())
    {
        rewriteAppendOnlyFileBackground();
    }
```

Reply:
@chenyang8094 i'm not sure just keeping this condition in serverCron is enough, we may need an additional fix, maybe in a separate PR.
I think that when the initial rewrite fails, we need to delete the manifest file and new incr file.
i.e. we have a manifest with only incr file, and no base, and i think that if we fail again and again we might have more and more incr files.
i'm not sure what'll happen if / when the rewrite eventually succeeds (how many incr files we'll have? we should have just the last).
could you maybe write a test that reproduce it, see what you get and solve it?

Reply:
@oranagra  This is the test I did (the method is to force exitFromChild(1)),

![image](https://user-images.githubusercontent.com/13696140/163770712-c128599f-01c5-4bff-8da0-4a17c7646a61.png)

![Uploading image.png…]()

Reply:
yes, they don't represent something user can recover from (it's an incomplete data set).
arguably, the manifest shouldn't even be placed as the main manifest file, but maybe in some temp name at that stage.
don't you agree?
what would happen if redis is then restarted? it'll try to load this file on startup?

the other concern i'm not sure about, is what happens when the rewrite eventually succeeds.
will we still have a base and several incr files, or just the last one?

Reply:
I don't think they need to be deleted. In my opinion, `config set appendonly yes `is just a special way to trigger AOFRW, which is essentially the same as the automatically triggered AOFRW. The user turns on the AOF, but he doesn't see any files (including manifest, etc.) on disk, which is very confusing.

The purpose of `config set appendonly yes`  is to enable AOF. The only difference is that there will be an `AOF_WAIT_REWRITE` state waiting for the first AOFRW to end. Although we generate three INCR AOFs, but they are empty and the user will not be aware of their presence when redis starts (redis will skip them and no log will be printed).

> the other concern i'm not sure about, is what happens when the rewrite eventually succeeds.
will we still have a base and several incr files, or just the last one?

To be sure, we'll end up with a BASE and an INCR。

Reply:
> The user turns on the AOF, but he doesn't see any files (including manifest, etc.) on disk, which is very confusing.

this was the state in all previous versions of redis, i don't see why it's wrong or confusing.

> `config set appendonly yes` is just a special way to trigger AOFRW

I'm probably taking it the wrong way, but i wanna stress that the `AOF_WAIT_REWRITE` state is very different than a normal AOFRW, redis already agreed to persist its data, and haven't done that yet, which is why it'll also refuse shutting down.

> they both generate multiple INCR AOFs

in one of these states, all these incr files are empty, so why keep them.

> The probability of initial AOFRW failure is very low, because before executing AOFRW, we will open the file to ensure that the system is normal

I disagree.. we know the disk is writable, but maybe there's insufficient memory for COW.
considering this could be the the first ever rewrite, i think the chances it'll fail are higher than a random re-occurring rewrite.

---

I think that as soon as someone disabled `appendonly` and then re-enabled it, the contents of the old AOF file is no longer valid or interesting (it'll be deleted anyway when the rewrite completes), we can either delete it right away, or maybe better "rename" it or store it in the manifest as history (so some people can use it for manual recovery in case the there's a failure at that time).
I think that keeping this old data presenting it (in the manifest) as valid data, is wrong.

Reply:
Before redis 7.0，If you `config set appendonly yes` you will see a file named "appendonly.aof"  be created immediately, even the first AOFRW keeps failing and keeps retrying.

> I think that as soon as someone disabled appendonly and then re-enabled it, the contents of the old AOF file is no longer valid or interesting (it'll be deleted anyway when the rewrite completes), we can either delete it right away, or maybe better "rename" it or store it in the manifest as history (so some people can use it for manual recovery in case the there's a failure at that time).
I think that keeping this old data presenting it (in the manifest) as valid data, is wrong.


I don't think those old AOFs can be deleted until AOFRW is really successful. The premise of AOF being deleted is that we have the correct AOFRW, so that we can restore our data from the newly generated base file.

Reply:
setting `appendonly no` doesn't mean we can delete them, but setting it back to `yes` means the old values are irrelevant.
i agree that it's a good idea not to delete them just yet. but on the other hand, keeping them in the same place is wrong too (if we'll crash, we'll try to recover from them. unlike what would have happened if we had crashed while `appendonly` is still set to `no`)

> Before redis 7.0，If you config set appendonly yes you will see a file named "appendonly.aof" be created immediately, even the first AOFRW keeps failing and keeps retrying.

that file is empty, right? so this means that as soon as you re-enabled `appendonly` it deleted the old content. is that right?

Reply:
> that file is empty, right? so this means that as soon as you re-enabled appendonly it deleted the old content. is that right?

No , this is the code before redis 7.0, redis use the `O_APPEND|O_CREAT` flag to open an existing AOF file.
```
int startAppendOnly(void) {
    char cwd[MAXPATHLEN]; /* Current working dir path for error messages. */
    int newfd;

    newfd = open(server.aof_filename,O_WRONLY|O_APPEND|O_CREAT,0644);
    serverAssert(server.aof_state == AOF_OFF);
    if (newfd == -1) {
        char *cwdp = getcwd(cwd,MAXPATHLEN);

        serverLog(LL_WARNING,
            "Redis needs to enable the AOF but can't open the "
            "append only file %s (in server root dir %s): %s",
            server.aof_filename,
            cwdp ? cwdp : "unknown",
            strerror(errno));
        return C_ERR;
    }
```

Reply:
ok. thanks (sorry for being lazy / busy).

so that means that in previous versions, after disable and re-enable AOF, if there's a crash in the initial rewrite, redis would have loading the old file when restarting (unlike a crash that would happen while it's still disabled).
don't you think that's wrong?

Reply:
@oranagra 
I understand what you care about, that is, we have three empty INCR files. I think we can solve it in the following way. When we open the next INCR AOF file, we first check whether the current AOF file is empty. If so, I will reuse the current AOF, WDYT?

Reply:
i actually don't mind the empty incr files, as long as their creation rate is limited.

what i mind is the odd state, in which you enabled AOF, it is not yet actually enabled (initial rewrite didn't complete or is failing in a loop), but the state on the disk is odd (either has an empty file, which i prefer to be missing, or has an old file from the previous time AOF was enabled, which i'm not sure should be there).

@yoav-steinberg @yossigo WDYT about this case
(this post should have enough state, no need to read the long discussion, or call me for details)

Reply:
1. In my opinion, `config set appendonly yes ` returns `OK`, the user will think that AOF is successfully enabled, although this is an asynchronous process in nature (even waiting for the first AOFRW)
2. If we change this behavior, I think it's a big breaking change
3. AOF is an important guarantee for data persistence. If the user mistakenly closes AOF in the middle, and then he realizes that he wants to re-open it again, just at this time AOFRW has been failing, then redis will face the risk of losing all data

Reply:
@chenyang8094 maybe i got confused again, we discussed empty incr files, but they shouldn't actually be empty.
i.e. if you enable AOF, the incr file keeps all the writes we revived during rewrite.
it is really wrong to put it in a manifest before we have a base file.
this is true for both a server on which AOF wasn't previously enabled (in which case we have an incr file without a base), and also the case where we do have a previous base file (in which case we have a time/data gap between the old AOF content, and the temporary incremental file, and we must never load it like as such).

regarding whether or not to delete the the previous AOF file when it was disabled, i wanna stress that it could be that all the data in the server got replaced after that (for instance changing assigned slot range, or replicating from a different master).
In fact, when a replica performs full-sync, before loading the rdb it calls `stopAppendonly`, and when it completed the rdb loading it calls `startAppendonly`, it could be replicating from a completely different master, and I argue that the AOF file must not be used for recovery (even if we crash before calling startAppendonly).
I wanna point out that in this state, the `server.aof_state` is `AOF_OFF`, but the `server.aof_enabled` config could be `1`.
This means that unlike the case where the user disabled appendonly and later re-enabled it, in this case if redis is restarted after stopAppendonly, the appendonly config at startup is still `on`

Reply:
I seems to me that when disabling AOF (`stopAppendonly`) we should always make sure to delete the manifest file (or at least rename it). We never want to load outdated data after disabling AOF.

Reply:
@yoav-steinberg Disable AOF does not mean that the previous persistent data is no longer important or outdated. 

Look at what we are doing in `stopAppendOnly`, we flush the AOF buffer to make sure it is persistent, this is to recover the data as much as possible when redis crashes after we disable AOF.

What I mean is that there are many reasons to disable AOF, and there is even the possibility of misoperation. We can't help users to assume that the previous AOF is no longer needed. On the contrary, I don't think there is any problem with our current logic, which is also the logic that has always existed before, maybe someone depends on this behavior.

```
void stopAppendOnly(void) {
    serverAssert(server.aof_state != AOF_OFF);
    flushAppendOnlyFile(1);
    redis_fsync(server.aof_fd);
    close(server.aof_fd);

    server.aof_fd = -1;
    server.aof_selected_db = -1;
    server.aof_state = AOF_OFF;
    server.aof_rewrite_scheduled = 0;
    killAppendOnlyChild();
}
```

Reply:
> it is really wrong to put it in a manifest before we have a base file.

If you care about this (though currently it doesn't pose any problems, doesn't it?), I think we can create this INCR lazily in `backgroundRewriteDoneHandler` (while in AOF_WAIT_REWRITE state). 

But as I said, if the user `config set appendonly yes` and see `OK`, but he doesn't see any AOF files  on disk（OAFRW is in endless failed retries） , this is inconsistent with the previous behavior.

Reply:
> > it is really wrong to put it in a manifest before we have a base file.

> If you care about this (though currently it doesn't pose any problems, doesn't it?), I think we can create this INCR lazily in backgroundRewriteDoneHandler (while in AOF_WAIT_REWRITE state).

that's a response to the top part of this message: https://github.com/redis/redis/pull/10582#discussion_r852914602 not sure you got my point.

In the previous AOF design, when appendonly is enabled, this data was buffered in RAM so it didn't mess up the AOF file on disk, but now it does, and i think that's a new bug.
it does post a problem since a restart after a crash during rewrite will either see a just an incr file with no base (loading it would be wrong), or if there was a pre-existing base, it'll see a base and a couple or incr files, but there's a time/data gap between the incr files (the executing the commands in the last incr file cold cause them to mess up the data, or at least to create a database that's not representing any point in time.

Reply:
@oranagra oh, I realize now that you're right, I'll make a separate PR where we'll continue the discussion on the resolution of this bug. thanks.